theano/tensor/opt.py

```diff
@@ -155,13 +155,16 @@ def broadcast_like(value, template, fgraph, dtype=None):
     if template not in fgraph.variables:
         raise NotImplementedError('broadcast_like currently requires the '
                                   'template Variable to be in the fgraph already')
+    if dtype is None:
+        dtype = template.dtype
+    value = T.cast(value, dtype)
+    if value.type == template.type:
```

Comment:
this do like on line 152. It would prevent the creation of the alloc when it isn't needed.